// @vuesion/service/dist/generators/module/index.js
"use strict";
const path = require("path");
const ast_1 = require("../ast");
const models_1 = require("@vuesion/models");
const path_1 = require("@vuesion/utils/dist/path");
const fileSystem_1 = require("@vuesion/utils/dist/fileSystem");
module.exports = {
    description: 'Add a module with VueX store and routes',
    prompts: [
        {
            type: 'input',
            name: 'name',
            message: 'What should it be called?',
            validate: (value) => {
                if (!value || value.length === 0) {
                    return 'name is required';
                }
                return fileSystem_1.folderExists(path_1.runtimeRoot(path.join(models_1.VuesionConfig.generators.outputDirectory, value)))
                    ? `folder already exists (${value})`
                    : true;
            },
        },
        {
            type: 'confirm',
            name: 'wantRoutes',
            default: true,
            message: 'Do you want routes?',
        },
        {
            type: 'confirm',
            name: 'wantVuex',
            default: true,
            message: 'Do you want vuex?',
        },
    ],
    actions: (data) => {
        const filePath = data.name.split('/');
        data.moduleName = filePath.pop();
        data.componentName = data.moduleName;
        data.basePath = path.join(process.cwd(), models_1.VuesionConfig.generators.outputDirectory, filePath.join('/'));
        let actions = [
            {
                type: 'add',
                path: '{{basePath}}/{{camelCase moduleName}}/{{properCase componentName}}/{{properCase componentName}}.vue',
                templateFile: path.join(process.cwd(), models_1.VuesionConfig.generators.blueprintDirectory, 'connected/connected.vue.hbs'),
                abortOnFail: true,
            },
            {
                type: 'add',
                path: '{{basePath}}/{{camelCase moduleName}}/{{properCase componentName}}/{{properCase componentName}}.spec.ts',
                templateFile: path.join(process.cwd(), models_1.VuesionConfig.generators.blueprintDirectory, 'connected/connected.spec.ts.hbs'),
                abortOnFail: true,
            },
        ];
        if (data.wantRoutes) {
            actions.push({
                type: 'add',
                path: '{{basePath}}/{{camelCase moduleName}}/routes.ts',
                templateFile: path.join(process.cwd(), models_1.VuesionConfig.generators.blueprintDirectory, 'module/routes.ts.hbs'),
                abortOnFail: true,
            });
            ast_1.addModuleToRouter(path.join(path.resolve(process.cwd()), models_1.VuesionConfig.generators.routerFile), data.moduleName);
        }
        if (data.wantVuex) {
            actions = actions.concat([
                {
                    type: 'add',
                    path: '{{basePath}}/{{camelCase moduleName}}/actions.spec.ts',
                    templateFile: path.join(process.cwd(), models_1.VuesionConfig.generators.blueprintDirectory, 'module/actions.spec.ts.hbs'),
                    abortOnFail: true,
                },
                {
                    type: 'add',
                    path: '{{basePath}}/{{camelCase moduleName}}/actions.ts',
                    templateFile: path.join(process.cwd(), models_1.VuesionConfig.generators.blueprintDirectory, 'module/actions.ts.hbs'),
                    abortOnFail: true,
                },
                {
                    type: 'add',
                    path: '{{basePath}}/{{camelCase moduleName}}/getters.spec.ts',
                    templateFile: path.join(process.cwd(), models_1.VuesionConfig.generators.blueprintDirectory, 'module/getters.spec.ts.hbs'),
                    abortOnFail: true,
                },
                {
                    type: 'add',
                    path: '{{basePath}}/{{camelCase moduleName}}/getters.ts',
                    templateFile: path.join(process.cwd(), models_1.VuesionConfig.generators.blueprintDirectory, 'module/getters.ts.hbs'),
                    abortOnFail: true,
                },
                {
                    type: 'add',
                    path: '{{basePath}}/{{camelCase moduleName}}/module.ts',
                    templateFile: path.join(process.cwd(), models_1.VuesionConfig.generators.blueprintDirectory, 'module/module.ts.hbs'),
                    abortOnFail: true,
                },
                {
                    type: 'add',
                    path: '{{basePath}}/{{camelCase moduleName}}/mutations.spec.ts',
                    templateFile: path.join(process.cwd(), models_1.VuesionConfig.generators.blueprintDirectory, 'module/mutations.spec.ts.hbs'),
                    abortOnFail: true,
                },
                {
                    type: 'add',
                    path: '{{basePath}}/{{camelCase moduleName}}/mutations.ts',
                    templateFile: path.join(process.cwd(), models_1.VuesionConfig.generators.blueprintDirectory, 'module/mutations.ts.hbs'),
                    abortOnFail: true,
                },
                {
                    type: 'add',
                    path: '{{basePath}}/{{camelCase moduleName}}/state.ts',
                    templateFile: path.join(process.cwd(), models_1.VuesionConfig.generators.blueprintDirectory, 'module/state.ts.hbs'),
                    abortOnFail: true,
                },
            ]);
            ast_1.addModuleToState(path.join(path.resolve(process.cwd()), models_1.VuesionConfig.generators.stateFile), data.moduleName);
        }
        return actions;
    },
};
